examples/segmentation_3d/assets/algo/algo.py

```diff
@@ -0,0 +1,78 @@
+import monai
+from monai.data import sliding_window_inference
+
+import torch
+
+import substratools as tools
+
+
+class MonaiAlgo(tools.algo.Algo):
+    device = torch.device("cpu")
+    loss_function = monai.losses.DiceLoss(do_sigmoid=True)
+
+    def _get_model(self):
+        model = monai.networks.nets.UNet(
+            dimensions=3,
+            in_channels=1,
+            out_channels=1,
+            channels=(16, 32, 64, 128, 256),
+            strides=(2, 2, 2, 2),
+            num_res_units=2,
+        ).to(self.device)
+        optimizer = torch.optim.Adam(model.parameters(), 1e-3)
+        return (model, optimizer)
+
+    def train(self, X, y, models, rank):
```

Comment:
It seems here, that the labels are encoded inside each x item.

This is a bit of a hack: it means that you have access in the predict function to the associated labels.

Reply:
Monai provides a DatasetLoader that randomizes input data and groups it in batches. This loader is an iterator over the batches. Because of the randomization, we cannot load X and y separately, otherwise matching values in X and y could be at different indexes and in different batches.

But the train and eval step use different loading strategy: randomized batches of 5 items for train, and non-randomized batches of 1 item for eval. My solution was therefore to have 2 openers.

For train, the loader would be returned by get_X (get_y returning None) and the train method would have to iterate over the X args to get the actual X and y values.

For test however, I made 2 generators that iterate of the loader, one for get_X and one for get_y. Since the loader is non-randomized, each instance of the loader returns the data in the same order. By adding generators on top of the loader that drop in one case the X value and in the other one the y value, we're able to load X and y separately.

To make sure that the train opener cannot be used for predictions, its save_prediction method raise an exception. This way any testtuple referencing this opener will fail.

You could however use the test opener for training, but it would quite slow as there is no batching and data is loaded twice, with no benefit.